### ISSUE TITLE
yolov5-lite

### ISSUE THREAD

**Issue by debapriyamaji** [ultralytics/yolov5#3168]
## 🚀 Feature
Yolov5 lite models: Making yolov5 more embedded friendly

## Motivation
 In line with the efficientdet-lite models that are more embedded friendly compared to efficientdet, is there any similar plan for yolov5-lite models. 

## Pitch
Following layers in the yolov5 are not embedded friendly:
- slice layer in the beginning.
- SiLU activation function.

If we can make suitable changes to these layers, these models can be deployed much more efficiently in embedded devices. 

I have trained some models with similar changes for the above mentioned layers and got accuracy within 2% of the original model.

If interested, I would love to share those results.

## Additional context
Efficientdet-lite models as compared to efficientdet.

**Comment by glenn-jocher** [comment]
@debapriyamaji sounds interesting, can you share some quantitative results before and after the changes? We have an activations study here that compares SiLU against some alternatives: https://wandb.ai/glenn-jocher/activations?workspace=user-glenn-jocher

I also recently created some documentation on the Focus() layer here that might interest you in https://github.com/ultralytics/yolov5/discussions/3181

**Comment by debapriyamaji** [comment]
@glenn-jocher Thanks for the reply and the insights. My ReLU results were in line with yours.

Following are the changes that I have made step by step for **yolov5s**:


Model config | PyTorch   mAP/AP50 |  GFLOPS | Comment
-- | --|--|--
Official model | 36.7/55.4 |17.0
Retrained Official model  | 36.9/55.6 | 17.0
SiLU replaced by ReLU | 34.9/53.7 | 17.0
SiLU replaced by ReLU + slice replaced by conv (Would like to call it **yolov5-lite**) | 35.0/54.4 | 17.1  | Replaced the initial slice layer with a conv(No=12,Ni=3, K=3, S=2). 

Let me know what you think.

**Comment by wudashuo** [comment]
I've done the exact same thing and have moved the model to devices months ago, following somebody's blog. The result I tested on devices seems to be okay, but it still has losses compared to yolov5s, besides, there are a few losses during migration.
As many devices don't support SiLU and Focus, I suppose many people would have done these changes to migrate the model to embedded devices, but there is still a lot to improve, not just simply replace SiLU with ReLU, replace Focus with Conv, that's just a compromise way. I'm working on it for months, trying to find out a model that less than 15 GFLOPs and more than 35 mAP. If you have some ideas, please let me know.
By the way, I don't think it should be called yolv5-lite, the parameters and GFLOPs are larger than yolov5s, it's not lighter than yolov5s.

**Comment by glenn-jocher** [comment]
@debapriyamaji @wudashuo one other point is that the Focus() module can be implemented with no slicing by allowing it to use the Contract() module. This provides no benefit on PyTorch/CUDA but may help other deployment targets.
https://github.com/ultralytics/yolov5/blob/b7cd1f540d5815b0a1cf2e23ce82a5fdb8f6b525/models/common.py#L163-L187

Also I'd argue that SiLU is well supported on many backends. We use it in our YOLOv5 CoreML models in our iOS App, where YOLOv5s runs in <18ms on iPhone 11/12. See https://apps.apple.com/app/id1452689527

**Comment by debapriyamaji** [comment]
Hi,
@wudashuo You can achieve your goal of 15GFLOPS and mAP of 35 by running **yolov5s6 at a resolution of 576x576**. Running inference on the pretrained checkpoint with i/p resolution of 576x576,  **I got  mAP of 37.7 for 14.15 GFLOPS**.

Regarding, yolov5-lite, GFLOPS number indeed look higher. However, these number are for convolution only. If you consider the operation required for sigmoid and multiplications needed for SiLU,  yolov5 will be higher than yolov5-lite. From, the context of efficientnet-lite, complexity is almost same as efficientnet. Main difference is in porting these models to embedded devices.

**Comment by debapriyamaji** [comment]
@glenn-jocher Thanks for the insight regarding Focus and SiLU layer. I have some follow-up questions:

- Is there any FPS benchmarking data available for different yolov5 models in apple devices?
- Did you run these models in 8 bit mode?
- How did you come to the conclusion that SiLU is well supported in apple devices? Did you run models with ReLU and SiLU and didn't observe any improvement in performance?

Thanks in advance.

**Comment by glenn-jocher** [comment]
@debapriyamaji see iOS iDetection Speed Table https://github.com/ultralytics/yolov5/issues/1276 for YOLOv5 benchmarks on iPhone models. Quantization seems to have no effect on ANE throughput. SiLU tests work perfectly well in iDetection.

**Comment by debapriyamaji** [comment]
@glenn-jocher Thanks for pointing to the FPS table. Is there any similar table for accuracy that would show the drop in accuracy because of quantization? 
I am trying to benchmark the ReLU model against SiLU model after quantization in tflite.  Will share the results once I am done . Since ReLU is more quantization frindly than SiLU, I am expecting similar trend here as well. Thanks.

**Comment by glenn-jocher** [comment]
@debapriyamaji there is no drop in accuracy when moving from FP32 to FP16 inference.

**Comment by debapriyamaji** [comment]
@glenn-jocher Sorry for not mentioning it in the previous post. I meant the drop in accuracy because of **FP8** since the CoreML models are exported as FP8. 

One more point I wanted to ask: A14 processor is ~11.0 TOPS. Yolov5s @320x192 is ~ **2.55GOPS**. If it runs at 14.3mS or ~**70FPS**, GOPS utilization is (70*2.55) GOPS = **178.5 GOPS**. That's only **1.6%** of the entire compute power.  So, the efficiency is quite low right?

Thanks.

**Comment by developer0hye** [comment]
@glenn-jocher 
Wow.. FReLU is so powerful... Do you have a plan to use FReLU instead of ReLU layer for new yolov5?

**Comment by glenn-jocher** [comment]
@developer0hye we have an activations study on YOLOv5s here:
https://wandb.ai/glenn-jocher/activations
<img width="608" alt="Screenshot 2021-06-10 at 12 03 06" src="https://user-images.githubusercontent.com/26833433/121506250-d8621d80-c9e3-11eb-9e81-0453710230f7.png">

Yes, FReLU performed very well, but you have to be careful interpreting these results as it's introducing additional convolutions into the model, which will especially help small models, but will cause faster overfitting in large models. Also memory usage increased from 12G to 18G when moving from SiLU to FReLU (though training speed was unaffected).

All in all it requires a lot more study into applying it to larger models, and perhaps only applying it to certain areas of the model rather than the entire model (could use feedback from the FReLU authors here), we just don't have time or manpower. 

EDIT1: I think memory usage could be reduced by not applying to as much to the earlier layers. In general the P0, P1, P2 layers in the backbone (first 1/3 of the backbone) all have very small strides with high memory usage and inference times.

**Comment by glenn-jocher** [comment]
@AyushExel can I borrow you to pass a feature request to the W&B team? We really need some semi-transparency on these legend overlays. Maybe 20-30% transparency would allow you to see the data under the legend, which would be great in my above screenshot.

**Comment by AyushExel** [comment]
@glenn-jocher I've passed it on :)

**Comment by glenn-jocher** [comment]
@AyushExel awesome thanks :) !

**Comment by Alex-afka** [comment]
@Glenn-jocher
Did you replace all the activation functions with frelu？
How does frelu work in yolov5l？

**Comment by glenn-jocher** [comment]
@Alex-afka see https://github.com/ultralytics/yolov5/issues/2891 for details

**Comment by PrashantDixit0** [comment]
@glenn-jocher and @debapriyamaji, Is this YOLOv5-lite pretrained models are avaialble for research or testing purposes ?

**Comment by glenn-jocher** [comment]
@PrashantDixit0 Yes, the YOLOv5s, YOLOv5m, YOLOv5l, YOLOv5x models are available with pretrained weights, for various tasks like object detection, instance segmentation, and more. You can find more information on using these models for research and testing purposes in the Ultralytics documentation at https://docs.ultralytics.com/yolov5/.
